core/src/main/scala/kafka/log/Log.scala

```diff
@@ -246,8 +246,11 @@ class Log(@volatile var dir: File,
     // The earliest leader epoch may not be flushed during a hard failure. Recover it here.
     _leaderEpochCache.clearAndFlushEarliest(logStartOffset)
 
-    assert(producerStateManager.isEmpty, "Producer state must be empty during log initialization")
-    loadProducerState(logEndOffset, reloadFromCleanShutdown = hasCleanShutdownFile, producerStateManager)
+    // Any segment loading or recovery code must not use producerStateManager, so that we can build the full state here
+    // from scratch.
+    if (!producerStateManager.isEmpty)
+      throw new IllegalStateException("Producer state must be empty during log initialization")
+    rebuildProducerState(logEndOffset, reloadFromCleanShutdown = hasCleanShutdownFile, producerStateManager)
```

Comment:
Here we want to call `loadProducerState`. This may be the cause of the test failures?

Reply:
Ah, didn't see the update. Guess the test failures are caused by something else.

Reply:
Yeah, the failure was essentially because we now call `updateFirstUnstableOffset` unconditionally (it used to be conditional on whether we are in an upgrade scenario or not). I don't think there is any harm in calling this either way, so updated the mocks in failing tests to expect a call to this.